var/spack/repos/builtin/packages/fastqc/package.py

```diff
@@ -0,0 +1,58 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+from distutils.dir_util import copy_tree, mkpath
+from distutils.file_util import copy_file
+
+
+class Fastqc(Package):
+    """A quality control tool for high throughput sequence data."""
+
+    homepage = "http://www.bioinformatics.babraham.ac.uk/projects/fastqc/"
+    url = "http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.5.zip"
+
+    version('0.11.5', '3524f101c0ab0bae77c7595983170a76')
+
+    depends_on('jdk')
+    depends_on('perl')          # for fastqc "script", any perl will do
+
+    patch('fastqc.patch', level=0)
+
+    def install(self, spec, prefix):
+        mkpath(self.prefix.bin)
+        mkpath(self.prefix.lib)
+        copy_file('fastqc', self.prefix.bin)
+        for j in ['cisd-jhdf5.jar', 'jbzip2-0.9.jar', 'sam-1.103.jar']:
+            copy_file(j, self.prefix.lib)
+        for d in ['Configuration', 'net', 'org', 'Templates', 'uk']:
+            copy_tree(d, join_path(self.prefix.lib, d))
+        chmod = which('chmod')
+        chmod('+x', join_path(self.prefix.bin, 'fastqc'))
+
+    def setup_environment(self, spack_env, env):
+        """Add <prefix> to the path; the package has a script at the
+           top level.
+        """
+        env.prepend_path('PATH', join_path(self.spec['jdk'].prefix, 'bin'))
```

Comment:
[So](https://www.fastcompany.com/3029762/how-a-popular-two-letter-word-is-undermining-your-credibility)...

If I do not have the autoload bit in my `etc/spack/modules.yaml` I end up with this *lmod* file:

```
whatis([[Name : fastqc]])
whatis([[Version : 0.11.5]])
help([[A quality control tool for high throughput sequence data.]])
prepend_path("PATH", "/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/fastqc-0.11.5-f66vjuaaao4s7r7nogz3r43h5dzojh27/bin")
prepend_path("CMAKE_PREFIX_PATH", "/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/fastqc-0.11.5-f66vjuaaao4s7r7nogz3r43h5dzojh27/")
prepend_path("LIBRARY_PATH", "/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/fastqc-0.11.5-f66vjuaaao4s7r7nogz3r43h5dzojh27/lib")
prepend_path("LD_LIBRARY_PATH", "/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/fastqc-0.11.5-f66vjuaaao4s7r7nogz3r43h5dzojh27/lib")
setenv("JAVA_HOME", "/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/jdk-8u92-linux-x64-4au5kjgku47oaoawsw7pes7hf6f32lnm")
```

If I make my etc/spack/modules.yaml file look like this:

```yaml
modules:
  enable::
    - lmod
  lmod:
    ^jdk:
      autoload: 'direct'
``` 

(adding the bottom three lines) then I get a module file that has these lines inserted after the `help` and before the `prepend_path("PATH...` lines.

```
if not isloaded("jdk/8u92-linux-x64-4au5kjg") then
    LmodMessage("Autoloading jdk/8u92-linux-x64-4au5kjg")
    load("jdk/8u92-linux-x64-4au5kjg")
end

if not isloaded("perl/5.24.0-ucc6rwm") then
    LmodMessage("Autoloading perl/5.24.0-ucc6rwm")
    load("perl/5.24.0-ucc6rwm")
end

```

That would cause the jdk module to be automagically loaded.  Oddly enough, it will **also** cause the perl module to be automagically loaded, even though I never mentioned perl in my modules.yaml.

Is that how it is supposed to work?